File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0
    const bridge = Array.from({length:bridge_length}).fill(0)
    let total_weight = 0
    
    while (truck_weights.length > 0 || total_weight > 0){
        // 다리에 올라가있는 트럭들의 무게의 합 + 다음 트럭의 무게의합 이 weight 보다 낮다면 트럭이 1초마다 출발한다.
        const arrived = bridge.shift()
        if (arrived !== 0){
            total_weight -= arrived
        }
        if (weight >= truck_weights[0] + total_weight){
            const poped = truck_weights.shift()
            bridge.push(poped)
            total_weight += poped
        }else {
            bridge.push(0)
        }
        
        time++
    }
    return time
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    // 남은 트럭 있냐  없으면 종료
    // 다리위 무게와 다음 트럭 무개가 가능하냐
        // 안되면 초만 흐름
        // 되면 올림
    
    int time = 0;
    int total_weight = 0;
    vector<int> arr(bridge_length,0); 
    while (true){
        time++;
        
        total_weight -= arr[0];
        arr.erase(arr.begin());
        
        // 
        if (!truck_weights.empty() && truck_weights[0] + total_weight <= weight){
            total_weight += truck_weights[0];
            arr.push_back(truck_weights[0]);
            truck_weights.erase(truck_weights.begin());
        } else {
            arr.push_back(0);
        }
        // 남은 트럭 없고, 다리 위 무게도 0이면 종료
        if (truck_weights.empty() && total_weight == 0) {
            return time;
        }
    }
    return time;
}